Prepare multi modality support

philippotto opened the pull request:
### Summary
- Supports rendering a dataset where layers can be transformed each (with an affine matrix).
- Transformation can be specified via datasource-properties.json or via JS API (will be ephemeral, then)
- Rendering was re-architectured to use cuckoo hashing for the dynamic look up of bucket addresses.
- Progressive rendering is done on the GPU (instead of on the CPU).
- For performance reasons, the plane geometries are now tessellated and the vertices are mutated by the vertex shader so that they are bucket-aligned. This enables us to precompute the cuckoo lookup in the vertex shader.
- The pre-computation is disabled when interpolation is enabled (would cause problems at chunk borders, otherwise). Also as a workaround for issues at chunk borders, pre-computation is ignored there, too (independent of interpolation).
- Transformed layers won't be rendered using the above optimization (since their buckets won't align with the vertices).
- Since enabling interpolation can incur a performance cost, it's disabled by default (won't affect old preferences) and the user will get a small hint in the left sidebar if it's enabled.
- As a consequence of the individually transformed layers, it can happen that one layer is rendered in a different magnification than another layer. This used to be uniform, but isn't anymore now (even though, it should still be uniform when no transformations are used).

### URL of deployed dev instance (used for testing):
- https://multimodality.webknossos.xyz

### To do:
- [X] allow to define transforms per layer which are used in the shader
- [X] adapt bucket picking to respect layer transformations
- [X] allow to make the active mag dependent on each layer (one layer can be rendered in mag 1 and the other in mag 4)
- [X] remove anchor point mechanism in favor of hashing to address buckets on the GPU
- [X] Backend: Store matrix from datasource-properties.json in postgres, include in dataset json sent to frontend
- Restore old behavior/functionality:
  - [X] progressive rendering
  - [x] data that is still being loaded should be gray instead of black
  - [x] "The currently rendered quality is not optimal due to the available magnifications and the viewport arrangement." shows false positively
  - [x] find my data needs to be adapted.
- Fix bugs:
  - [x] having more layers in a DS than can be rendered at the same time results in rendering issues
  - [x] rotation from arbitrary mode is not removed when switching back to orthogonal mode
  - [x] deep freeze breaks for transform matrices apparently
  - [X] thin lines at bucket borders (solved by disabling anti aliasing which prevents flat varyings to be interpolated errorenously)
  - [X] data offset by 32vx
  - [X] missing data at border
  - [X] weird triangles at border
  - [x] sometimes bucket strip shows at viewport boundary shows wrong data
  - [x] some transformations cause incorrect data to be rendered (needs investigation)
  - [x] XZ and YZ rendering is corrupt (probably the computation in the vertex shader is not adapted yet)
  - [x] sometimes cross hairs are invisible (probably due to disabled antialiasing)
  - [X] support segmentation layers again
  - [x] fix line artifact when rendering segmentation layer (this is caused by the segmentation patterns which rely on partial derivatives; unfortunately, these seem to be buggy now that there's a special tessellation)
- [X] Performance (solved by performing bucket look up in vertex shader)
- [ ] clean up UI
  - [x] display correct mag etc
  - [x] fix CI tests
  - [x] fix screenshot tests
  - [x] add warning for performance for interpolation?
  - [x] disable performance optimization when interpolation is enabled
  - [x] change default for interpolation
  - [ ] doublecheck bucket loading order for transformed layers
- [x] go through to do comments in code

### Steps to test:
- test view mode and annotation mode for multiple datasets
- pay attention to performance and rendering issues

### Issues:
- fixes #6085

------
(Please delete unneeded items, merge only when none are left open)
- [x] Updated [(unreleased) changelog](../blob/master/CHANGELOG.unreleased.md#unreleased)
- [ ] Updated [(unreleased) migration guide](../blob/master/MIGRATIONS.unreleased.md#unreleased) if applicable
- [ ] Updated [documentation](../blob/master/docs) if applicable
- [ ] Adapted [wk-connect](https://github.com/scalableminds/webknossos-connect) if datastore API changes
- [ ] Adapted [wk-libs python client](https://github.com/scalableminds/webknossos-libs/tree/master/webknossos/webknossos/client) if relevant API parts change
- [ ] Needs datastore update after deployment
- [x] Ready for review

philippotto commented on frontend/javascripts/oxalis/model_initialization.ts:
```diff
@@ -222,6 +224,122 @@ export async function initialize(
     setInitialTool();
   }
 
+  const makeTranslation = (x: number, y: number, z: number): Matrix4x4 =>
```
move into own module?

philippotto replied:
I moved it into the front-end api.

philippotto commented on frontend/javascripts/oxalis/shaders/segmentation.glsl.ts:
```diff
@@ -273,12 +262,15 @@ export const getSegmentationId: ShaderModule = {
           worldPositionUVW,
-          vec4(0.0, 0.0, 0.0, 0.0)
+          vec4(0.0, 0.0, 0.0, 0.0),
+          // todo
```
todo

philippotto commented on frontend/javascripts/oxalis/shaders/texture_access.glsl.ts:
```diff
@@ -37,6 +39,15 @@ export const getRgbaAtIndex: ShaderModule = {
         ).rgba;
     }
+
+    uvec4 getUnsignedRgbaAtIndex(highp usampler2D dtexture, float textureWidth, float idx) {
```
use overloading instead of separate name?

philippotto commented on frontend/javascripts/oxalis/shaders/texture_access.glsl.ts:
```diff
@@ -104,56 +200,66 @@ export const getColorForCoords: ShaderModule = {
-        // rendering wrong data.
-        returnValue[1] = vec4(1.0, 1.0, 0.0, 1.0);
+      if (worldPositionUVW.x < 0. || worldPositionUVW.y < 0. || worldPositionUVW.z < 0.) {
+        // should alpha be -1.0?
```
todo

philippotto commented on frontend/javascripts/oxalis/geometries/plane.ts:
```diff
@@ -4,20 +4,23 @@ import { getBaseVoxelFactors } from "oxalis/model/scaleinfo";
   OrthoViewValues,
 } from "oxalis/constants";
 
+export const PLANE_SUBDIVISION = 200;
```
use in shader

daniel-wer replied:
Maybe you could also add a comment on what the rationale behind this number is. It's probably not important that it's exactly 200, but explaining why it isn't 20 or 2000 might help future readers :)

philippotto replied:
good point :) I just thought the value through (the initial value was more a gut feeling) and decreased it to 80 now (and added the explanation of course).

philippotto commented on frontend/javascripts/oxalis/shaders/main_data_fragment.glsl.ts:
```diff
@@ -251,5 +282,254 @@ void main() {
+
+  const uint FALLBACK_COUNT = 3u;
+
+  // todo: seg layer is missing!
```
todo

philippotto commented on frontend/javascripts/oxalis/shaders/main_data_fragment.glsl.ts:
```diff
@@ -251,5 +282,254 @@ void main() {
+
+  gl_Position = projectionMatrix * modelViewMatrix * vec4(position, 1.0);
+
+  float planeWidth = 376.;
+  float subdivisionCount = 200.;
```
dont use hardcoded values

philippotto commented on frontend/javascripts/oxalis/model/accessors/flycam_accessor.ts:
```diff
@@ -312,32 +364,35 @@ export function getValidTaskZoomRange(
 export function isMagRestrictionViolated(state: OxalisState): boolean {
   const { resolutionRestrictions } = state.tracing.restrictions;
-  const zoomStep = getRequestLogZoomStep(state);
+  // todo: the resolution restrictions need to be stored per layer (backend + frontend).
```
@fm3 How would you gauge the necessary effort to store the resolution restrictions per layer? I think, this PR should be merged soon and a follow-up one could tackle this, but I'd still be interested in your thoughts.

philippotto replied:
for the record: this only affects task annotations. for now, I added a heuristic which uses the first color layer to read the active mag. for layers without transforms this doesn't matter, anyway.

daniel-wer commented on frontend/javascripts/oxalis/geometries/materials/plane_material_factory.ts:
```diff
@@ -455,6 +489,35 @@ class PlaneMaterialFactory {
+            }
+          }
+
+          // If all layers have a transform, the representativeLayerIdxForMag
```
Could you briefly explain what the `representativeLayerIdxForMag` is needed/used for?

philippotto replied:
Good point! I added the following comment:

> // The vertex shader looks up the buckets for rendering so that the
// fragment shader doesn't need to do so. Currently, this only works
// for layers that don't have a transformation (otherwise, the differing
// grids wouldn't align with each other).
// To align the vertices with the buckets, the current magnification is
// needed. Since the current mag can differ from layer to layer, the shader
// needs to know which mag is safe to use.
// For this purpose, we define the representativeLayerIdxForMag which is
// essentially an index to *some* layer that doesn't have any transforms.
// If all layers have a transform, the representativeLayerIdxForMag
// isn't relevant which is why it can default to 0.

daniel-wer commented on frontend/javascripts/oxalis/geometries/materials/plane_material_factory.ts:
```diff
@@ -697,17 +763,18 @@ class PlaneMaterialFactory {
     // layers were least-recently activated (but are now disabled).
     // The first array contains the color layer names and the second the segmentation layer names.
```
What does the third return value represent? From the code it looks like it is the overall number of layers, but then, shouldn't that also be returned in line 766?

philippotto replied:
I think the case where `0` is returned doesn't really matter as nothing will be rendered, anyway, but it's probably less confusing to return the global layer count. I'll adapt the code (and add a comment) :)

daniel-wer commented on frontend/javascripts/oxalis/geometries/materials/plane_material_factory.ts:
```diff
@@ -759,17 +826,25 @@ class PlaneMaterialFactory {
+  getFragmentShaderWithUniforms(): [string, Uniforms] {
     const { initializedGpuFactor, maximumLayerCountToRender } =
       Store.getState().temporaryConfiguration.gpuSetup;
     // Don't compile code for segmentation in arbitrary mode
```
This comment no longer seems relevant (even before your change) :thinking:

daniel-wer commented on frontend/javascripts/oxalis/geometries/materials/plane_material_factory.ts:
```diff
@@ -780,24 +855,36 @@ class PlaneMaterialFactory {
-}`;
+    const { initializedGpuFactor, maximumLayerCountToRender } =
+      Store.getState().temporaryConfiguration.gpuSetup;
+    // Don't compile code for segmentation in arbitrary mode
```
See my comment above.

daniel-wer commented on frontend/javascripts/oxalis/geometries/materials/plane_material_factory.ts:
```diff
@@ -780,24 +855,36 @@ class PlaneMaterialFactory {
+      segmentationLayerNames,
+      packingDegreeLookup,
+      // Todo: this is not computed per layer. See #4018
+      dataTextureCountPerLayer: Model.maximumTextureCountForLayer,
```
In https://github.com/scalableminds/webknossos/issues/4018 you mention that this could improve performance if it was "fixed". Just wanted to mention it in case you had forgotten about it (like me), no need to do it in this PR.

daniel-wer commented on frontend/javascripts/oxalis/model/bucket_data_handling/abstract_cuckoo_table.ts:
```diff
@@ -0,0 +1,303 @@
+  }
+
+  checkValidKey(key: K) {
+    if (key === EMPTY_KEY) {
```
There are a couple of usages of `EMPTY_KEY` in this class. Shouldn't they use `getEmptyKey()`?

daniel-wer commented on frontend/javascripts/oxalis/model/bucket_data_handling/abstract_cuckoo_table.ts:
```diff
@@ -0,0 +1,303 @@
+
+  private initializeTableArray() {
+    this.table = new Uint32Array(AbstractCuckooTable.ELEMENTS_PER_ENTRY * this.entryCapacity).fill(
+      EMPTY_KEY_VALUE,
```
I'm stumbling about this line, since a subclass could decide not to use `EMPTY_KEY_VALUE` and instead return something else in `getEmptyValue()`. If you use that instead, there should be a check whether `ELEMENTS_PER_ENTRY > 1` and if so that all entries of the empty key are the same, before using `getEmptyValue()[0]` to fill the array.

philippotto replied:
You are right! I guess, my generalization of these classes got a bit too flexibel (which wouldn't have worked if the subclasses were to push the limits). I cleaned this up (mainly by making the restrictions clearer in the comments + interface). See 7e4db9587.

daniel-wer reviewed (commented):
First part of my review :) I'll continue tomorrow.

philippotto commented on frontend/javascripts/oxalis/geometries/materials/plane_material_factory.ts:
```diff
@@ -232,6 +226,14 @@ class PlaneMaterialFactory {
+      const layer = getLayerByName(Store.getState().dataset, dataLayer.name);
+
+      this.uniforms[`${layerName}_transform`] = {
+        value: invertAndTranspose(layer.transformMatrix || Identity4x4),
```
I'm not 100% sure what you mean, but it might be correct :smile: I added a comment to the `invertAndTranspose` function which explains its existence. Copied here so that you don't need to look it up:

> // Transposition is often needed so that the matrix has the right format
// for matrix operations (e.g., on the GPU; but not for ThreeJS).
// Inversion is needed when the position of an "output voxel" (e.g., during
// rendering in the fragment shader) needs to be mapped to its original
// data position (i.e., how it's stored without the transformation).
// Without the inversion, the matrix maps from stored position to the position
// where it should be rendered.

daniel-wer commented on frontend/javascripts/oxalis/shaders/texture_access.glsl.ts:
```diff
@@ -86,16 +88,85 @@ export const getColorForCoords: ShaderModule = {
+      // be precise because the triangles of the plane won't necessarily
+      // align with the bucket borders.
+      <% if (isFragment) { %>
+        if (supportsPrecomputedBucketAddress) {
```
I did not find an explanation for this parameter. I expect it to be related to the triangle border artifacts you experienced in an intermediate version. Is that correct?

philippotto replied:
`supportsPrecomputedBucketAddress` is true if a layer doesn't have special transforms and if interpolation is not used. Only in these cases, the bucket address that was precomputed in the vertex shader can be used.

daniel-wer commented on frontend/javascripts/oxalis/shaders/texture_access.glsl.ts:
```diff
@@ -104,56 +175,66 @@ export const getColorForCoords: ShaderModule = {
+      uint renderedMagIdx;
+
+      if (!supportsPrecomputedBucketAddress) {
+        for (uint i = 0u; i < 4u; i++) {
```
Is the `< 4` coming from the maximum allowed zoom step diff that is still rendered? If so, I'd reuse the respective constant via templating.

philippotto commented on frontend/javascripts/oxalis/controller/scene_controller.ts:
```diff
@@ -450,13 +465,21 @@ class SceneController {
+    this.layerBoundingBoxes.forEach((bbCube) => {
+      const visible = id !== OrthoViews.TDView || tdViewDisplayLayerBorders;
+      bbCube.setVisibility(visible);
+      if (visible) {
```
good point! I just adapted the visibility of the bboxes to the layers and this was indeed a bug :+1: (however, the old code at least respected the tdViewDisplayLayerBorders setting successfully for some reason).

daniel-wer commented on frontend/javascripts/oxalis/shaders/main_data_fragment.glsl.ts:
```diff
@@ -134,15 +154,28 @@ ${compileShader(
-    gl_FragColor = vec4(0.0, 0.0, 0.0, 1.0);
-    return;
+
+  if (floor(flatVertexPos.x) == floor(worldCoordUVW.x) && floor(flatVertexPos.y) == floor(worldCoordUVW.y)) {
+    gl_FragColor = vec4(1., 0., 1., 1.);
+    // return;
+  }
+
+  if (index.x == 200.) {
+      gl_FragColor = vec4(0., 1., 1., 1.);
+//      return;
```
Looks like remains from testing that can be removed

daniel-wer commented on frontend/javascripts/oxalis/shaders/main_data_fragment.glsl.ts:
```diff
@@ -251,5 +284,249 @@ void main() {
+
+uniform float activeMagIndices[<%= globalLayerCount %>];
+uniform uint availableLayerIndexToGlobalLayerIndex[<%= globalLayerCount %>];
+uniform vec3 resolutions[20];
```
The 20 probably shouldn't be hard-coded

daniel-wer commented on frontend/javascripts/oxalis/shaders/main_data_fragment.glsl.ts:
```diff
@@ -251,5 +284,249 @@ void main() {
+  });
+}
+
+export function getMainVertexShader(params: Params) {
```
The name of the file suggests that it only contains a fragment shader. It should be adapted.

daniel-wer commented on frontend/javascripts/oxalis/shaders/main_data_fragment.glsl.ts:
```diff
@@ -251,5 +284,249 @@ void main() {
+const float bucketWidth = <%= bucketWidth %>;
+const float bucketSize = <%= bucketSize %>;
+const float l_texture_width = <%= l_texture_width %>;
+
```
Does the above section mirror the whole variable declarations of the fragment shader?

philippotto replied:
good point! I could dry most of it :)

daniel-wer commented on frontend/javascripts/oxalis/shaders/main_data_fragment.glsl.ts:
```diff
@@ -251,5 +284,249 @@ void main() {
+  //   ^ first vertex
+  //          ^ second vertex
+  //                   ...             ^ last vertex
+  //                             ^ if a is the second last vertex, it needs to be moved to the right
```
Doesn't the same issue exist for the third last vertex then, sometimes. That it maybe would need to be moved to the right, as otherwise the second last bucket would not be covered? (and so on...)

daniel-wer commented on frontend/javascripts/oxalis/shaders/main_data_fragment.glsl.ts:
```diff
@@ -251,5 +284,249 @@ void main() {
+  //                               instead of the left.
+
+  // Calculate the index of the vertex (e.g., index.x=0 is the first horizontal vertex).
+  index = (position.xy / (PLANE_WIDTH / 2.) + 1.) / 2. * PLANE_SUBDIVISION;
```
Does it work to use the `VIEWPORT_WIDTH` constant for that? Don't the orthogonal viewport possibly have wildly varying sizes?
Also, I'm failing to follow the calculation I have to admit :see_no_evil:

philippotto replied:
I added a better comment to the code. I hope this makes it clearer :) See de948413f

philippotto commented on frontend/javascripts/oxalis/controller/combinations/skeleton_handlers.ts:
```diff
@@ -268,7 +268,8 @@ function addNode(
       rotation,
       OrthoViewToNumber[Store.getState().viewModeData.plane.activeViewport],
-      getRequestLogZoomStep(Store.getState()),
+      // todo: this would be a breaking change. clarify.
```
see https://scm.slack.com/archives/C5AKLAV0B/p1677079138825069 for discussion.

daniel-wer commented on frontend/javascripts/oxalis/shaders/main_data_fragment.glsl.ts:
```diff
@@ -251,5 +284,249 @@ void main() {
+  index = (position.xy / (PLANE_WIDTH / 2.) + 1.) / 2. * PLANE_SUBDIVISION;
+
+  uint activeMagIdx = uint(activeMagIndices[representativeLayerIdxForMag]);
+  vec2 d = transDim(vec3(32.) * getResolution(activeMagIdx)).xy;
```
Could you clarify what `d` is? Maybe we should have a look at this code together if it's easier :)

daniel-wer commented on frontend/javascripts/oxalis/shaders/main_data_fragment.glsl.ts:
```diff
@@ -251,5 +284,249 @@ void main() {
+
+  flatVertexPos = worldCoordUVW;
+  float NOT_YET_COMMITTED_VALUE = pow(2., 21.) - 1.;
+  const float bucketWidth = 32.;
```
I'd suggest to use `BUCKET_WIDTH` via templating here as well. Same for the FALLBACK_COUNT in line 489.

daniel-wer reviewed (commented):
I've gone through all the files once, now :) Very impressive work! Most of the changes were really easy to follow :+1: 

With testing, I'd wait until the first round of feedback is addressed and most ToDos have been taken care of.

philippotto commented on frontend/javascripts/oxalis/controller/combinations/move_handlers.ts:
```diff
@@ -45,9 +47,10 @@ export const moveW = (deltaW: number, oneSlide: boolean): void => {
   if (oneSlide) {
-    const logZoomStep = getRequestLogZoomStep(Store.getState());
+    const activeMagIndices = getActiveMagIndicesForLayers(Store.getState());
+    const finestMagIdx = _.min(values(activeMagIndices)) || 0;
     const wDim = Dimensions.getIndices(activeViewport)[2];
-    const wStep = getResolutions(Store.getState().dataset)[logZoomStep][wDim];
+    const wStep = getResolutions(Store.getState().dataset)[finestMagIdx][wDim];
```
> On a different note, maybe it's more intuitive if this only takes into account layers that are currently visible? Otherwise, if the layer with the finest resolution is not visible, the step might be too small and there will be no change in the viewports.

This is a good point :+1: 

> I'm not sure I understand how this continues to work 
My current assumption is that there are layers with varying scales and, therefore, the one with the finest resolution not necessarily is the one with the smallest voxel size.
Instead, it's likely the case that the resolutions refer to a base scale - in which case it does make sense again.

Let's discuss this in-person :)

philippotto commented on frontend/javascripts/oxalis/model_initialization.ts:
```diff
@@ -222,6 +224,122 @@ export async function initialize(
+    setTransforms(layerName, makeRotation(theta, pos));
+  };
+
+  type Spec =
```
I put them into the frontend api for now and marked them as experimental (I also slimmed it down a bit). Hope that's okay!

philippotto commented on frontend/javascripts/test/puppeteer/dataset_rendering.screenshot.ts:
```diff
@@ -1,3 +1,4 @@
+import "test/mocks/lz4";
```
Good question :see_no_evil: I assume that the import dependencies changed somehow so that the lz4 dependency is imported now (which fails on node). I didn't bother to investigate further. Hope this is alright!

philippotto commented on frontend/javascripts/oxalis/model/reducers/settings_reducer.ts:
```diff
@@ -178,9 +179,14 @@ function SettingsReducer(state: OxalisState, action: Action): OxalisState {
+          return newState;
+        }
+        // Restore rotation
+        return setRotationReducer(newState, [0, 0, 0]);
```
The rotation is only changed if the user does so in flight/oblique mode. Since this affects the matrix (which is also used in orthogonal mode), the rotation needs to be reset. Does this make sense? I also extended the comment a bit.

philippotto commented on frontend/javascripts/oxalis/model/bucket_data_handling/layer_rendering_manager.ts:
```diff
@@ -105,21 +92,42 @@ function consumeBucketsFromArrayBuffer(
+  optSanitizer?: (arg: string) => string,
+): number {
+  const sanitizer = optSanitizer || _.identity;
+  const allColorLayers = getColorLayers(Store.getState().dataset);
```
good point! I don't know why I implemented it so convoluted. thanks :)

philippotto commented on frontend/javascripts/oxalis/model/bucket_data_handling/layer_rendering_manager.ts:
```diff
@@ -175,58 +186,56 @@ export default class LayerRenderingManager {
-    const isArbitrary = constants.MODES_ARBITRARY.includes(viewMode);
     const { sphericalCapRadius } = state.userConfiguration;
     const isVisible = isLayerVisible(dataset, this.name, datasetConfiguration, viewMode);
+    const rects = getViewportRects(Store.getState());
```
I think, the "areas" are a concept which are a bit outdated now. They basically contain top, left, right, bottom in bucket coordinates. This means that the areas didn't necessarily change when one moved only by one voxel (since the same buckets were at the border of the viewport). With the transformation of layers, this doesn't really make much sense anymore, though. The oblique bucket picker needs the exact extent of the rendered viewport to calculate its things (one could probably use a similar optimization technique, though, but I don't feel like this would get us much).

The areas are only used by the prefetching saga now. This could use some overhauling, too, but obviously it's out of scope for this PR.

philippotto commented on frontend/javascripts/oxalis/model/bucket_data_handling/texture_bucket_manager.ts:
```diff
@@ -3,40 +3,41 @@ import _ from "lodash";
-// The first value f denotes the following:
-//   If f >= 0, f denotes the index in the data texture where the bucket is stored.
-//   If f == -1, the bucket is not yet committed
-//   If f == -2, the bucket is not supposed to be rendered. Out of bounds.
```
The cuckoo table now accomplishes these cases as follows:
f >= 0 ==> [key, address] is in table
f == -1 ==> [key, NOT_YET_COMMITTED_VALUE] is in table
f == -2 ==> key doesn't exist in table

philippotto commented on frontend/javascripts/oxalis/model/bucket_data_handling/bucket_picker_strategies/oblique_bucket_picker.ts:
```diff
@@ -3,103 +3,162 @@ import type { EnqueueFunction } from "oxalis/model/bucket_data_handling/layer_re
+    // of horizontal lines which cover the entire rendered plane.
+    // These "scan lines" are traversed to find out which buckets need to be
+    // sent to the GPU.
+    const zDiff = 10;
```
The coordinates are essentially in "screen space" and multiplied with the `queryMatrix`. So, a static value makes somewhat sense in my opinion. However, I'm sure this could be improved somehow. However, since I'm not sure how, I'd leave it as is for the moment :smile:

philippotto commented on frontend/javascripts/oxalis/model/bucket_data_handling/layer_rendering_manager.ts:
```diff
@@ -45,19 +44,7 @@ const asyncBucketPick: typeof asyncBucketPickRaw = memoizeOne(
-
-  return map3((_pos, idx) => roundToNearestBucketBoundary(position, idx), position);
-}
+const getLookUpCuckooTable = memoizeOne(() => new CuckooTableVec5(256));
```
Yes, it's a lazy-initialized singleton. I renamed it to `getSharedLookUpCuckooTable` and added a comment about it being a lazy singleton.

daniel-wer commented on frontend/javascripts/oxalis/shaders/main_data_shaders.glsl.ts:
```diff
@@ -470,8 +415,11 @@ void main() {
+  debugVal = getWorldCoordUVW();
+
+  gl_Position = projectionMatrix * viewMatrix * worldCoord;
+  gl_Position.z = originalZ;
```
Maybe you could add a small comment here, explaining that due to floating point inaccuracies z might change unintendedly, otherwise

daniel-wer commented on frontend/javascripts/oxalis/shaders/segmentation.glsl.ts:
```diff
@@ -271,12 +271,18 @@ export const getSegmentationId: ShaderModule = {
+      vec4[2] volume_color;
+      vec3 transformedCoordUVW = transDim((<%= segmentationName %>_transform * vec4(transDim(worldPositionUVW), 1.0)).xyz);
+      if (isOutsideOfBoundingBox(transformedCoordUVW)) {
+        return volume_color;
```
`volume_color` is uninitialized at this point, is the returned value deterministic in that case or should we initialize it to be sure?

philippotto replied:
good question! just tested it and it should be safe :)

daniel-wer commented on frontend/javascripts/test/model/binary/data_rendering_logic.spec.ts:
```diff
@@ -1,10 +1,11 @@
-// @ts-nocheck
```
:tada:

daniel-wer commented on frontend/javascripts/test/puppeteer/dataset_rendering.screenshot.ts:
```diff
@@ -137,13 +152,18 @@ const datasetConfigOverrides: Record<string, PartialDatasetConfiguration> = {
   testFn: () => Promise<boolean>,
   resolveFn: (arg0: boolean) => void,
 ) {
+  // todo: revert
```
Todo comment

daniel-wer commented on frontend/javascripts/oxalis/controller/combinations/move_handlers.ts:
```diff
@@ -48,6 +48,9 @@ export const moveW = (deltaW: number, oneSlide: boolean): void => {
 
   if (oneSlide) {
     const activeMagIndices = getActiveMagIndicesForLayers(Store.getState());
+    // The following logic might not always make sense when having layers
```
Another candidate for the followup (limitation documentation) issue

daniel-wer reviewed (commented):
Looking very good, please see my additional comments :)

- [x] Quite a few of the screenshots are missing the cross hair in one or more of the data viewports. I think this was fixed in the mean time, so it's probably enough to re-generate the screenshots
- [x] What do you think about adding a transformed dataset to the screenshot tests?

If I remember correctly, the team already did some monkey testing. What do you think remains to be tested or should we do a final monkey testing round in the end?

philippotto commented on frontend/javascripts/oxalis/controller/scene_controller.ts:
```diff
@@ -474,12 +475,13 @@ class SceneController {
+    Object.keys(this.layerBoundingBoxes).forEach((layerName) => {
+      const bbCube = this.layerBoundingBoxes[layerName];
+      const visible =
+        id === OrthoViews.TDView && tdViewDisplayLayerBorders && !layerNameToIsDisabled[layerName];
```
Yeah, I think, I went that route because the new layer bboxes are yellow now (to have at least a bit of a distinguishing characteristic) and if the layer bbox == dataset bbox, the borders would be yellow in the data viewports which felt a bit dominant.

philippotto commented on frontend/javascripts/oxalis/model/accessors/flycam_accessor.ts:
```diff
@@ -552,3 +566,34 @@ function _getUnrenderableLayerInfosForCurrentZoom(
   _getUnrenderableLayerInfosForCurrentZoom,
 );
+
+function _getActiveResolutionInfo(state: OxalisState) {
```
Both callers of this function (status bar and info tab view) make proper use of the return values :)

philippotto commented on frontend/javascripts/oxalis/model/accessors/flycam_accessor.ts:
```diff
@@ -372,33 +374,46 @@ export function getValidTaskZoomRange(
 ): [number, number] {
   const defaultRange = [baseDatasetViewConfiguration.zoom.minimum, Infinity] as Vector2;
   const { resolutionRestrictions } = state.tracing.restrictions;
+  // We use the first color layer as a heuristic to check the validity of the zoom range,
+  // as we don't know to which layer a restriction is meant to be applied.
+  // If the layers don't have any transforms, the layer choice doesn't matter, anyway.
```
Good point! I opened https://github.com/scalableminds/webknossos/issues/6926.

philippotto replied:
Good point! I opened https://github.com/scalableminds/webknossos/issues/6926.

philippotto commented:
@daniel-wer I incorporated your remaining feedback and also updated the screenshots. I'll add an additional screenshot test for the transformations in a follow-up PR. If you accept the PR, I'll merge it today :)

daniel-wer commented on frontend/javascripts/oxalis/api/api_latest.ts:
```diff
@@ -1970,9 +1970,14 @@ class DataApi {
 
   /*
-   * _Experimental_ API for creating transformation matrices based on an array of TransformerSpecs.
+   * _Experimental_ API for creating transformation matrices based on an array of TransformerSpec's.
```
```suggestion
   * _Experimental_ API for creating transformation matrices based on an array of TransformerSpecs.
```

daniel-wer reviewed (approved):
LGTM :tada:
